lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1424 @@
+          });
+  cache.insert({type, value});
+  return value;
+}
+
+/// Construct a null value of the given type using the given builder.
+static Value createNullValue(ImplicitLocOpBuilder &builder, FIRRTLType type) {
+  SmallDenseMap<FIRRTLType, Value> cache;
+  return createNullValue(builder, type, cache);
+}
+
+/// Helper function that inserts reset multiplexer into all `ConnectOp`s and
+/// `PartialConnectOp`s with the given target. Looks through `SubfieldOp`,
+/// `SubindexOp`, and `SubaccessOp`, and inserts multiplexers into connects to
+/// these subaccesses as well. Modifies the insertion location of the builder.
+/// Returns true if the `resetValue` was used in any way, false otherwise.
+static bool insertResetMux(ImplicitLocOpBuilder &builder, Value target,
+                           Value reset, Value resetValue) {
+  // Indicates whether the `resetValue` was assigned to in some way. We use this
+  // to erase unused subfield/subindex/subaccess ops on the reset value if they
+  // end up unused.
+  bool resetValueUsed = false;
+
+  for (auto &use : target.getUses()) {
+    Operation *useOp = use.getOwner();
+    builder.setInsertionPoint(useOp);
+    TypeSwitch<Operation *>(useOp)
+        // Insert a mux on the value connected to the target:
+        // connect(dst, src) -> connect(dst, mux(reset, resetValue, src))
+        .Case<ConnectOp, PartialConnectOp>([&](auto op) {
+          if (op.dest() != target)
+            return;
+          LLVM_DEBUG(llvm::dbgs() << "  - Insert mux into " << op << "\n");
+          auto muxOp = builder.create<MuxPrimOp>(reset, resetValue, op.src());
+          op.srcMutable().assign(muxOp);
+          resetValueUsed = true;
+        })
+        // Look through subfields.
+        .Case<SubfieldOp>([&](auto op) {
+          auto resetSubValue =
+              builder.create<SubfieldOp>(resetValue, op.fieldnameAttr());
+          if (insertResetMux(builder, op, reset, resetSubValue))
+            resetValueUsed = true;
+          else
+            resetSubValue.erase();
+        })
+        // Look through subindices.
+        .Case<SubindexOp>([&](auto op) {
+          auto resetSubValue =
+              builder.create<SubindexOp>(resetValue, op.indexAttr());
+          if (insertResetMux(builder, op, reset, resetSubValue))
+            resetValueUsed = true;
+          else
+            resetSubValue.erase();
+        })
+        // Look through subaccesses.
+        .Case<SubaccessOp>([&](auto op) {
+          auto resetSubValue =
+              builder.create<SubaccessOp>(resetValue, op.index());
+          if (insertResetMux(builder, op, reset, resetSubValue))
+            resetValueUsed = true;
+          else
+            resetSubValue.erase();
+        });
+  }
+  return resetValueUsed;
+}
+
+//===----------------------------------------------------------------------===//
+// Reset Network
+//===----------------------------------------------------------------------===//
+
```

Comment:
I would really appreciate a big block comment talking about the algorithm here.  It is a lot of stuff, but there is no "forest through the trees" on what is happening.

Reply:
Yeah good point. Let me add some details on what the pass does.

Reply:
Done.